src/core/column/qcut.h

```diff
@@ -0,0 +1,178 @@
+
+namespace dt {
+
+/**
+  *  Virtual column to bin numeric values into equal-population
+  *  discrete intervals, i.e. quantiles. In reality, for some data
+  *  these quantiles won't have exactly the same population.
+  *
+  *  Quantiles are generated based on the element/group information
+  *  obtained from the groupby operation, i.e. rowindex and offsets.
+  *  These groups, having ids 0, 1, ..., ngroups - 1, are binned
+  *  into `nquantiles` equal-width discrete intervals. As a result,
+  *  all the duplicates of a value `x` will go to the same `x_q` quantile.
+  *
+  *  In a simple case, when all the values are missing or constant,
+  *  `Qcut_ColumnImpl::make()` will immediately return a pointer to the
+  *  corresponding virtual column implementation: `ConstNa_ColumnImpl`
+  *  or `ConstInt_ColumnImpl`.
+  *
+  *  In the general case `Qcut_ColumnImpl::make()` will return a pointer
+  *  to the `Latent_ColumnImpl` that wraps a `Qcut_ColumnImpl*`.
+  *  This column implementation will be materialized on the first access
+  *  to the data, because it is much more efficient to generate
+  *  all the quantiles at once in parallel.
+  *
+  */
+class Qcut_ColumnImpl : public Virtual_ColumnImpl {
+  private:
+    Column col_;
+    RowIndex ri_;
+    Groupby gb_;
+    int32_t nquantiles_;
+    size_t : 32;
+
+  public:
+    static ColumnImpl* make(Column&& col, int32_t nquantiles) {
+      xassert(nquantiles > 0);
+      auto res = group({col}, {SortFlag::NONE});
+      RowIndex ri = std::move(res.first);
+      Groupby gb = std::move(res.second);
+
+      // If there is one group only, fill it with constants or NA's.
+      if (gb.size() == 1) {
+        if (col.get_element_isvalid(0)) {
+          return new ConstInt_ColumnImpl(
+                   col.nrows(),
+                   (nquantiles - 1) / 2,
+                   SType::INT32
+                 );
+        } else {
+          return new ConstNa_ColumnImpl(col.nrows(), SType::INT32);
+        }
+      }
+
+      return new Latent_ColumnImpl(new Qcut_ColumnImpl(
+                   std::move(col), nquantiles, std::move(ri), std::move(gb)
+                 ));
+
+    }
+
+
+    void materialize(Column& col_out, bool) override {
+      Column col_out_ = Column::new_data_column(col_.nrows(), SType::INT32);
+      auto data_out_ = static_cast<int32_t*>(col_out_.get_data_editable());
+
+      constexpr double epsilon = static_cast<double>(
+                                   std::numeric_limits<float>::epsilon()
+                                 );
+
+      // Check if there is an NA group.
+      bool has_na_group;
+      {
+        size_t row;
+        bool row_valid = ri_.get_element(0, &row);
+        xassert(row_valid); (void) row_valid;
+        has_na_group = !col_.get_element_isvalid(row);
+      }
+
+      // Set up number of valid groups and the quantile coefficient.
+      size_t ngroups = gb_.size() - has_na_group;
+      double a = nquantiles_ * (1 - epsilon) / (ngroups - 1);
```

Comment:
what if `ngroups` is 1 (after removing the na group)?

Reply:
Right, will handle this case separately.